### DISCUSSION TITLE
How to change a PDF's title?

### DISCUSSION THREAD

**Discussion by jon-m-craig** [LibrePDF/OpenPDF#1006] [Q&A] [answered]
So I have a template PDF (it's a medical claim form), with fields for everything that needs filled in.  This part is easy, I can easily fill in the required data from my application.  But even poring over the Javadocs I cannot figure out how to change the Stamper-generated PDF's title.  There is no setTitle() method or similar...anywhere.

**Comment by Lonzak** [comment]
What kind of title do you mean? You can set the PDF metadata if you want. There are two types of metadata: the DID (document information dictionary) and XMP metadata.

DID:
It is the `PdfStamper.setMoreInfo(...)` method you want to take a look at.

XMP (a bit more effort since you have to edit/create it by yourself. You might want to use adobe's [XMP lib](https://search.maven.org/artifact/com.adobe.xmp/xmpcore) for that):
`PdfStamper/PdfWriter setXmpMetadata(...)`

**Reply by jon-m-craig** [reply]
Good question - the title shown in this screen clip as "CMS 1500 Template" - I'd like to change that on the fly to something like "Invoice #10000" or whatever I need.
![20231218_07h52m37s_grim](https://github.com/LibrePDF/OpenPDF/assets/24698280/9578dfad-d1a8-4f06-bb37-f962302629e3)

**Reply by jon-m-craig** [reply]
So looking at the `moreInfo()` methods; they're deprecated.  The `infoDictionary()` methods are what we're supposed to use instead.  Both of the `get`s return nothing, so the title I'm needing to change is not in those.  Where is that "CMS-1500 Template" title shown in that screen cap coming from??!!?!

**Comment by Lonzak** [comment]
>Where is that "CMS-1500 Template" title shown in that screen cap coming from??!!?!

It is either the file name or it is the PDF title defined in the XMP metadata. Do you have *dobe Reader installed? Then open it and under File\Properties\Document Properties\Descripton (Tab) you'll find it. Or post the file here and we can check...

**Reply by jon-m-craig** [reply]
Yeah - it is the XMP metadata apparently, it is in the bytes returned by `reader.getMetadata()` - so now I have to figure out how to alter this.

**Reply by Lonzak** [reply]
As mentioned above I used *dobe's XML lib for that which makes it a bit more convenient...

**Reply by jon-m-craig** [reply]
Yeah, I've added that to my project -- any quick-n-dirty snippet/example of using it to get/set metadata using a `PdfStamper`/etc.  I'm not finding many great examples online.

**Reply by jon-m-craig** [reply]
In fact 'dobe only talks publicly about a C++ implementation; I can't find Java info anywhere.

**Reply by Lonzak** [reply]
> In fact 'dobe only talks publicly about a C++ implementation; I can't find Java info anywhere.

This is indeed another topic :-( 
https://github.com/adobe/XMP-Toolkit-SDK/issues/5

**Comment by Lonzak** [comment]
But I am still using version 5.X so it might need some changes if you use the latest version:    

     //reading
     public static XMPMeta getXMPMetaData(PdfReader pdfReader) throws SystemException {

        byte[] xmpMeta = null;
        try {
            xmpMeta = pdfReader.getMetadata();

            if (xmpMeta == null || xmpMeta.length == 0) {
                return null;
            }

            //set the parsing options to the most compatible/lax settings
            ParseOptions po = new ParseOptions();
            po.setAcceptLatin1(true);
            po.setDisallowDoctype(false);
            po.setFixControlChars(true);
            po.setOmitNormalization(true);
            po.setRequireXMPMeta(false);
            po.setStrictAliasing(false);

            return XMPMetaFactory.parseFromBuffer(xmpMeta,po);
        }
        catch (Exception e) {
            //Exceptionhandling
        }
    }

     //writing
    public static void setXMPMetaData(PdfStamper pdfstamper, final List<Namespaces> namespaces, final List<XMPMetaDTO> meta) throws Exception{

        // register namespaces
        XMPSchemaRegistry registry = XMPMetaFactory.getSchemaRegistry();
        Iterator<NamespaceDTO> iter = namespaces.iterator();
        while (iter.hasNext()) {
            NamespaceDTO act = iter.next();

            try {
                registry.registerNamespace(act.getNamespace(), act.getPrefix());
            }
            catch (XMPException e) {
                //handle
            }
        }

        // convert metadata to *dobe
        XMPMetaImpl xmpMeta = new XMPMetaImpl();
        Iterator<XMPMetaDTO> mIter = meta.iterator();
        while (mIter.hasNext()) {
            XMPMetaDTO act = mIter.next();
            try {
                xmpMeta.setProperty(act.getNamespace(), act.getElementName(), act.getElementValue());
            }
            catch (XMPException e) {
                //handle
            }
        }

        // serialize metadata and write it to the pdf
        byte[] serializedMeta;
        try {
            serializedMeta = XMPSerializerHelper.serializeToBuffer(xmpMeta, null);
        }
        catch (XMPException e) {
           //handle
        }

        pdfstamper.setXmpMetadata(serializedMeta);
    }

**Reply by jon-m-craig** [reply]
This is rather bizarre; I'm obviously missing something.

```
            XMPMetaImpl xmp = new XMPMetaImpl();
            xmp.setProperty("http://purl.org/dc/elements/1.1/", "title", "TITLE");
            xmp.setProperty("http://purl.org/dc/elements/1.1/", "creator", "CREATOR");
            byte[] newXMP = XMPSerializerHelper.serializeToBuffer(xmp, null);

            flatStamper.setXmpMetadata(newXMP);

            flatStamper.setFormFlattening(true);
            flatStamper.close();
```

Does not change the metadata... it still shows all the same as before.

**Reply by jon-m-craig** [reply]
Whoa - it gets weirder - if I call `getMetadata()` on the generated PDF, and log it, it is:

```
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta x:xmptk="Adobe XMP Core 6.1.11" xmlns:x="adobe:ns:meta/">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description dc:creator="CREATOR" dc:title="TITLE" rdf:about="" xmlns:dc="http://purl.org/dc/elements/1.1/"></rdf:Description>
  </rdf:RDF>
</x:xmpmeta>                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
<?xpacket end="w"?>
```

However viewing the PDF still shows the old title and hitting "Properties" still shows the old title/creator/etc.

Uhhh... what is happening here?

**Reply by Lonzak** [reply]
Yes doesn't look too bad.
You need to use correct namespace. I mean what is the result of your original file with the title from your screenshot? Read it and then try to create it... I created one with *dobe and it looks like this:

    <x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="Adobe XMP Core 9.1-c001 79.2a0d8d9, 2023/03/14-11:19:46        ">
	<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
		<rdf:Description rdf:about="" 
		xmlns:pdf="http://ns.adobe.com/pdf/1.3/" 
		xmlns:xmp="http://ns.adobe.com/xap/1.0/" 
		xmlns:xmpRights="http://ns.adobe.com/xap/1.0/rights/" 
		xmlns:dc="http://purl.org/dc/elements/1.1/" 
			<pdf:Producer>openPDF</pdf:Producer>
			<pdf:Keywords>keywords; </pdf:Keywords>
			<xmp:ModifyDate>2023-12-18T16:11:45+01:00</xmp:ModifyDate>
			<xmp:MetadataDate>2023-12-18T16:11:45+01:00</xmp:MetadataDate>
			<xmp:CreateDate>2023-12-18T16:11:45+01:00</xmp:CreateDate>
			<xmpRights:Marked>True</xmpRights:Marked>
			<xmpRights:WebStatement>copyright url  </xmpRights:WebStatement>
			<dc:format>application/pdf</dc:format>
			<dc:title>
				<rdf:Alt>
					<rdf:li xml:lang="x-default">document title  </rdf:li>
				</rdf:Alt>
			</dc:title>
			<dc:creator>
				<rdf:Seq>
					<rdf:li>author </rdf:li>
				</rdf:Seq>
			</dc:creator>
			<dc:description>
				<rdf:Alt>
					<rdf:li xml:lang="x-default">description  </rdf:li>
				</rdf:Alt>
			</dc:description>
			<dc:subject>
				<rdf:Bag>
					<rdf:li>keywords</rdf:li>
					<rdf:li></rdf:li>
				</rdf:Bag>
			</dc:subject>
		</rdf:Description>
	</rdf:RDF>
</x:xmpmeta>

**Reply by jon-m-craig** [reply]
The full metadata dump is, 

```
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta x:xmptk="Adobe XMP Core 6.1.11" xmlns:x="adobe:ns:meta/">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description adhocwf:state="1" adhocwf:version="1.1" dc:format="application/pdf" pdf:Keywords="" pdf:Producer="1500CMS.COM[[[000000" rdf:about="" xmlns:adhocwf="http://ns.adobe.com/AcrobatAdhocWorkflow/1.0/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:pdf="http://ns.adobe.com/pdf/1.3/" xmlns:xmp="http://ns.adobe.com/xap/1.0/" xmlns:xmpMM="http://ns.adobe.com/xap/1.0/mm/" xmp:CreateDate="2006-06-05T12:52:48Z" xmp:CreatorTool="" xmp:MetadataDate="2022-01-27T15:34:36-05:00" xmp:ModifyDate="2022-01-27T15:34:36-05:00" xmpMM:DocumentID="uuid:3a0604a5-6a7e-4ddd-8b13-5e56c5031c15" xmpMM:InstanceID="uuid:2b442251-ef6e-4451-b97d-cca6c32e0521">
      <dc:creator>
        <rdf:Seq>
          <rdf:li>1500CMS.COM</rdf:li>
        </rdf:Seq>
      </dc:creator>
      <dc:title>
        <rdf:Alt>
          <rdf:li xml:lang="x-default">CMS-1500 Template</rdf:li>
        </rdf:Alt>
      </dc:title>
      <dc:description>
        <rdf:Alt>
          <rdf:li xml:lang="x-default"></rdf:li>
        </rdf:Alt>
      </dc:description>
    <pdf:Producer>1500CMS.COM[[[000000; modified using OpenPDF 1.3.34</pdf:Producer><xmp:ModifyDate>2023-12-18T10:37:30-05:00</xmp:ModifyDate></rdf:Description>
  </rdf:RDF>
</x:xmpmeta>                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
<?xpacket end="w"?>
```

And that's AFTER the code that should have changed the title and creator.

**Reply by jon-m-craig** [reply]
It gets EVEN WEIRDER!  Here is a metadata dump, after getting it to write correctly:

```
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta x:xmptk="Adobe XMP Core 6.1.11" xmlns:x="adobe:ns:meta/">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description adhocwf:state="1" adhocwf:version="1.1" dc:format="application/pdf" pdf:Keywords="" pdf:Producer="MedFlex[[[000000" rdf:about="" xmlns:adhocwf="http://ns.adobe.com/AcrobatAdhocWorkflow/1.0/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:pdf="http://ns.adobe.com/pdf/1.3/" xmlns:xmp="http://ns.adobe.com/xap/1.0/" xmlns:xmpMM="http://ns.adobe.com/xap/1.0/mm/" xmp:CreateDate="2006-06-05T12:52:48Z" xmp:CreatorTool="" xmp:MetadataDate="2022-01-27T15:34:36-05:00" xmp:ModifyDate="2022-01-27T15:34:36-05:00" xmpMM:DocumentID="uuid:3a0604a5-6a7e-4ddd-8b13-5e56c5031c15" xmpMM:InstanceID="uuid:2b442251-ef6e-4451-b97d-cca6c32e0521">
      <dc:creator>
        <rdf:Seq>
          <rdf:li>MedFlex</rdf:li>
        </rdf:Seq>
      </dc:creator>
      <dc:title>
        <rdf:Alt>
          <rdf:li xml:lang="x-default">IT IS FINISHED</rdf:li>
        </rdf:Alt>
      </dc:title>
      <dc:description>
        <rdf:Alt>
          <rdf:li xml:lang="x-default"></rdf:li>
        </rdf:Alt>
      </dc:description>
    <pdf:Producer>1500CMS.COM[[[000000; modified using OpenPDF 1.3.34</pdf:Producer><xmp:ModifyDate>2023-12-18T11:02:31-05:00</xmp:ModifyDate></rdf:Description>
  </rdf:RDF>
</x:xmpmeta>                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
                                                                                                   
<?xpacket end="w"?>
```

However viewing this PDF, it shows...

![20231218_11h08m54s_grim](https://github.com/LibrePDF/OpenPDF/assets/24698280/5b9d5829-adab-4492-948e-ee19faf17810)

**Comment by Lonzak** [comment]
Ok that doesn't make sense. Sure you got the correct document (reloaded)?

**Reply by jon-m-craig** [reply, marked as answer]
AHA!  I figured it out.  Even though the `InfoDictionary` is empty in the Reader, creating a Stamper from the Reader moves the original XMP data into the corresponding keys in the `InfoDictionary` in the Stamper.  So setting the `InfoDictionary` in the Stamper... does finally work.  Apparently that overrides the XMP data in viewers.  VERY CONFUSING.
